# PYMONGO CLEAN MOVIES
***
# <center> Tunggal Manda Ary Triyono
# ---------------------------------------------------------------------------------------------------------------

# Import Library

In [1]:
import pymongo
from pymongo import MongoClient
import pprint
from  IPython.display import clear_output
import warnings
warnings.filterwarnings('ignore')

# Movies Collection 

In [2]:
client_mov = MongoClient('mongodb+srv://userstudent:admin1234@cluster0-nnbxe.gcp.mongodb.net/test?retryWrites=true&w=majority')

db_mov = client_mov['sample_mflix']
col_mov = db_mov.movies

print('Total Documents on Movies :', col_mov.count())

Total Documents on Movies : 23541


***
Mendefinisikan Variabel variabel yang dipakai pada proses Cleaning Movies_Initial menjadi clean_movies
<br>variabel `client_mov` untuk menginisialisasi cluster yang dipaai yakni cluster0-nnbxe
<br>variabel `db_mov` yang berisi nama database yakni database sample_mflix
<br>variabel `col_mov` yang berisi nama collection yakni movies
<br>
<br> Kemudian dapat dilihat bahwa Total Documents pada Movies : **23541**
***

# Movies Initial Collection

In [3]:
client_mov_init = MongoClient('mongodb+srv://adminuser12345:adminuser12345@cluster0-phmf4.gcp.mongodb.net/test?retryWrites=true&w=majority')

db_mov_init = client_mov_init['sample_mflix']
col_mov_init = db_mov_init.movies_initial

print('Total Documents on Movies Initial :', col_mov_init.count())

Total Documents on Movies Initial : 46014


***
Mendefinisikan Variabel variabel yang dipakai pada proses Cleaning Movies_Initial menjadi clean_movies
<br>variabel `client_mov_init` untuk menginisialisasi cluster yang dipaai yakni cluster0-phmf4
<br>variabel `db_mov_init` yang berisi nama database yakni database sample_mflix
<br>variabel `col_mov_init` yang berisi nama collection yakni movies
<br>
<br> Kemudian dapat dilihat bahwa Total Documents pada Movies Initial : **46014**
***

# Exploring The Data

# Different Movies VS Movies Initial

In [81]:
#Movies
pprint.pprint(list(col_mov.find({'title': 'The House of the Devil'})))

[{'_id': ObjectId('573a13c0f29313caabd62078'),
  'awards': {'nominations': 5, 'text': '3 wins & 5 nominations.', 'wins': 3},
  'cast': ['Jocelin Donahue', 'Tom Noonan', 'Mary Woronov', 'Greta Gerwig'],
  'countries': ['USA'],
  'directors': ['Ti West'],
  'fullplot': 'College student Samantha Hughes takes a strange babysitting job '
              'that coincides with a full lunar eclipse. She slowly realizes '
              'her clients harbor a terrifying secret',
  'genres': ['Horror'],
  'imdb': {'id': 1172994, 'rating': 6.4, 'votes': 26606},
  'languages': ['English'],
  'lastupdated': '2015-09-06 00:44:02.603000000',
  'metacritic': 73,
  'num_mflix_comments': 1,
  'plot': 'College student Samantha Hughes takes a strange babysitting job '
          'that coincides with a full lunar eclipse. She slowly realizes her '
          'clients harbor a terrifying secret',
  'poster': 'https://m.media-amazon.com/images/M/MV5BMTAxMDAxODg5ODReQTJeQWpwZ15BbWU3MDI5ODYxODI@._V1_SY1000_SX677_AL_.

In [82]:
#Movies Initial
pprint.pprint(list(col_mov_init.find({'title': 'The House of the Devil'})))


[{'_id': ObjectId('5db0085107645f129de54594'),
  'awards': '',
  'cast': "Jeanne d'Alcy, Georges M�li�s",
  'country': 'France',
  'director': 'Georges M�li�s',
  'fullplot': 'A bat flies into an ancient castle and transforms itself into '
              'Mephistopheles himself. Producing a cauldron, Mephistopheles '
              'conjures up a young girl and various supernatural creatures, '
              'one of which brandishes a crucifix in an effort to force the '
              'devil-vampire to vanish.',
  'genre': 'Short, Horror',
  'imdbID': 91,
  'imdbRating': 6.8,
  'imdbVotes': 1135,
  'language': '',
  'lastupdated': '2015-08-26 00:06:16.697000000',
  'metacritic': '',
  'plot': 'A bat flies into an ancient castle and transforms itself into '
          'Mephistopheles himself. Producing a cauldron, Mephistopheles '
          'conjures up a young girl and various supernatural creatures, one '
          '...',
  'poster': '',
  'rating': '',
  'released': '',
  'runtime': '3 

***
Pada data diatas dapat dilihat bahwa terdapat **perbedaan Data** antar documents Movies dan Movies Initial.
<br>Sebagian besar isi datanya berbeda baik **keys** dan **values** nya
<br>Untuk perbedaan keys dan values dapat lebih jelas dilihat pada code selanjutnya
***

# Keys Different Movies VS Movies Initial

In [27]:
#Movies
key_mov = [list(doc.keys()) for doc in col_mov.find({})]
key_mov = set([ j for i in key_mov for j in i])
key_mov

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

In [37]:
#Movies Initial
key_mov_init = [list(doc.keys()) for doc in col_mov_init.find({})]
key_mov_init = set([ j for i in key_mov_init for j in i])
key_mov_init

{'_id',
 'awards',
 'cast',
 'country',
 'director',
 'fullplot',
 'genre',
 'imdbID',
 'imdbRating',
 'imdbVotes',
 'language',
 'lastupdated',
 'metacritic',
 'plot',
 'poster',
 'rating',
 'released',
 'runtime',
 'title',
 'type',
 'writer',
 'year'}

In [43]:
print('TOTAL FIELDS MOVIES         :',len(key_mov),'\n--------------------------------')
print('TOTAL FIELDS MOVIES INITIAL :',len(key_mov_init))

TOTAL FIELDS MOVIES         : 22 
--------------------------------
TOTAL FIELDS MOVIES INITIAL : 22


***
Dapat dilihat bahwa Keys antara movies dan movies initial **jumlahnya sama** yakni **22** namun **jenis fieldnya berbeda**. Salah satu tantangan pada use case ini adalah untuk membuat collection dengan keys yang sama dengan keys pada collection Movies baik jumlah maupun jenis fieldnya menggunakan collection movies initial.
***

# Some Values Different Movies VS Movies Initial

### Keys : Awards

In [5]:
#Movies
awards_mov = list(col_mov.distinct('awards'))
awards_mov[:5]

[{'nominations': 0, 'text': '1 win.', 'wins': 1},
 {'nominations': 1, 'text': '1 nomination.', 'wins': 0},
 {'nominations': 2, 'text': '2 nominations.', 'wins': 0},
 {'nominations': 0, 'text': '3 wins.', 'wins': 3},
 {'nominations': 0, 'text': '2 wins.', 'wins': 2}]

In [8]:
#Movies Initial
awards_mov_init = list(col_mov_init.distinct('awards'))
print(awards_mov_init[:5],end='\n')

['', '1 win.', '2 wins.', '1 nomination.', '2 nominations.']


***
Dapat dilihat bahwa pada **Keys** : `awards`, terdapat perbedaan antara movies dan movies initial. Pada **movies** values berupa dictionary yang berisi **nomination, text, dan win**. Sedangkan pada **movies initial**, values berupa **string yang berbeda beda** tiap documentnya.
***

### Keys : Runtime

In [62]:
#Movies
list(col_mov.find({},{'_id':0, 'runtime': 1}).limit(1))

[{'runtime': 78}]

In [63]:
#Movies Initial
list(col_mov_init.find({},{'_id':0, 'runtime': 1}).limit(1))

[{'runtime': '1 min'}]

***
Pada Keys : **Runtime** juga terdapat perbedaan antara kedua collection tersebut. Pada movies initial terdapat kata min yang mengikuti jumlah menitnya, sehingga perlu dihilangkan Min nya dan hanya mengambil angka menitnya.
***

***
Sebagian Besar Keys lain juga formatya berbeda antara kedua collection tersebut. Hanya ada 9 Keys yang sama formatnya sehingga tidak perlu dilakukan perubahan format, yakni : fullplot, lastupdated, metacritic, plot, poster, rating, title, type, year.
***

# Keys for Equalize the collection

In [17]:
imdb = list(col_mov.find({},{'_id':0, 'imdb':1}))
imdb_list=[]
for i in imdb :
    imdb_list.append(list(i.values())[0])
print(len(imdb_list,'\n'))
print(imdb_list[:5])

[{'id': 6517, 'rating': 6.3, 'votes': 162},
 {'id': 8133, 'rating': 7.8, 'votes': 4680},
 {'id': 12512, 'rating': 6.8, 'votes': 489},
 {'id': 13025, 'rating': 7.8, 'votes': 3738},
 {'id': 15498, 'rating': 7.1, 'votes': 327}]

***
Untuk Membuat clean Movies yang isinya sama dengan movies, perlu dibuat semacam **primary key** yang sama antara movies initial dan movie. Pada Kode ini, dibuatlah suatu list yang berisi **kumpulan values dari keys imdb** dari **movies collection**. **List** ini **digunakan** untuk tahap selanjutnya dimana untuk **menyeleksi dokumen** pada tahap selanjutnya. Jadi data yang ada pada clean movies nantinya adalah data movies initial dengan values imdb yang sama dengan values imdb dari data movies.
***

# 1. Make  new collection named clean_movies_tunggal

In [61]:
pipeline = [
    {
        '$project': {
            ### this code is to include the keys for a new collection on database ###
            ### 1 for include , 0 for exclude the keys ###
            ### Split some keys into an array value, renaming as well ###         
            ### adding embeded documents ###
            ### grouping the several keys into embeded documents by one keys ###
            
            'awards': {'nominations': {'$arrayElemAt':
                                           [{'$split':
                                                 [{'$arrayElemAt':
                                                       [{'$split': [ "$awards", " nomination" ]}, -2]},' ']},
                                            -1]},
                       'text': "$awards",
                       'wins': {'$arrayElemAt':
                                    [{'$split':
                                          [{'$arrayElemAt':
                                                [{'$split': [ "$awards", " win" ]}, -2]},' ']},
                                     -1]}
                      },
            
            'casts' : {"$cond": [{'$eq': ["$cast", ""]}, 
                                          '$nonExistinField' , 
                                          {'$split': ["$cast", ", "]},]},
            
            'countries': {"$cond": [{'$eq': ["$country", ""]}, 
                                              '$nonExistinField' , 
                                              {'$split': ["$country", ", "]},]}, 
            
            'directors': {"$cond": [{'$eq': ["$director", ""]}, 
                                              '$nonExistinField' , 
                                              {'$split': ["$director", ", "]},]},
            
            'fullplot': {"$cond": [{'$eq': ["$fullplot", ""]}, 
                                              '$nonExistinField' , 
                                              "$fullplot",]},
            
            'genres': {"$cond": [{'$eq': ["$genre", ""]}, 
                                            '$nonExistinField' , 
                                            {'$split': ["$genre", ", "]},]},
            
            'imdb': {
                     'id': "$imdbID",
                     'rating': "$imdbRating",
                     'votes': "$imdbVotes"
                    },
            
            'language': {"$cond": [{'$eq': ["$language", ""]}, 
                                              '$nonExistinField' , 
                                              {'$split': ["$language", ", "]},]},
            
            'lastupdated': {"$cond": [{'$eq': ["$lastupdated", ""]}, 
                                                 '$nonExistinField' , 
                                                 '$lastupdated']},
            
            'metacritic': {"$cond": [{'$eq': ["$metacritic", ""]}, 
                                              '$nonExistinField' , 
                                              "$metacritic"]},
                       
            'num_mflix_comments': '',
            
            'plot': {"$cond": [{'$eq': ["$plot", ""]}, 
                                          '$nonExistinField' , 
                                          '$plot']},
            
            'poster': {"$cond": [{'$eq': ["$poster", ""]}, 
                                          '$nonExistinField' , 
                                          '$poster']},
            
            'rated':{"$cond": [{'$eq': ["$rating", ""]}, 
                                          '$nonExistinField' , 
                                          '$rating']},
            
            'released': {"$cond": [{'$eq': ["$released", ""]}, 
                                              '$nonExistinField' , 
                                              {'$dateFromString': {'dateString': "$released"}} 
                                  ]},
            
            'runtime': {"$cond": [{'$eq': ["$runtime", ""]}, 
                                  '$nonExistinField' ,
                                  {'$arrayElemAt':[{'$split': [ "$runtime", " " ]}, 0]}
                                  ]}, #100 min to 100
            
            'title': {"$cond": [{'$eq': ["$title", ""]}, 
                                          '$nonExistinField' , 
                                          '$title']},
            
            'tomatoes': {
                
                         'lastUpdated': "$lastupdated",
                         'viewer': {'rating': "$rating"}
                        },
            
            'type': {"$cond": [{'$eq': ["$type", ""]}, 
                                          '$nonExistinField' , 
                                          '$type']},
            
            'writers': {"$cond": [{'$eq': ["$type", ""]}, 
                                          '$nonExistinField' , 
                                          {'$split': ["$writer", ", "]}]},
                        
            'year': {"$cond": [{'$eq': ["$year", ""]}, 
                                          '$nonExistinField' , 
                                          '$year']}
        }
    },
    {
        ### matching imdb values between movies & movies initial ### 
       '$match': { "imdb": { '$in': [dict(sorted(imdb_list[x].items(), key=lambda x: x[0])) for x in range(len(imdb_list))]}}
    },
    {
        ### naming the new collection as output ### 
        '$out': "clean_movies_tunggal"
    }
]

clear_output()
pprint.pprint(list(col_mov_init.aggregate(pipeline)))

[]


***
Kode diatas untuk membuat collection baru dengan nama clean_movies_tunggal. Pada bagian **projecting**(\$project) fields yang didefinisikan adalah sesuai dengan fields-fields yang ada pada collection movies. Isi dari fields-fields tersebut didapat dari movies initial dengan format yang disesuaikan dengan movies.

>`awards` isinya adalah nomination, text dan win. misal untuk mengisi nilai nomination, idenya adalah mengambil awards dari yang bentuknya string, kemudian di split berdasarkan nomination dan dipilih hasil split yang sebelah kiri, kemudian displit lagi berdasarkan spasi dan data yang akan diambil adalah nomor 1 dari belakang. kemudian untuk field text adalah isi dari string awards pada movies initial, dan untuk wins sistemnya sama seperti nomination naun di split berdasarkan win.

>`imdb` isinya id, rating, votes yang masing masing didapatkan dari field imdbID, imdbRating dan imdbVotes dari collection movies initial

>`tomatoes` isinya lastupdate dan viewer yang isinya rating. lastupdate dan rating asing masing didapatkan dari field lastupdate dan rating dari collection movies initial

>`fullplot`, `genres`, `lastupdated`, `metacritic`, `plot`, `poster`, `rated`, `title`, `type`, `writers`, `year` hanya ditampilkan sesuai dengan format values pada movies initial

>`cast`, `countries`, `directors`, `language`, `runtime` isinya adalah values pada field field tersebut pada movies initial yang di split berdasarkan koma(,)

>`released` isinya adalah datetime yang berbentuk datetime. Caranya adalah dengan mengubah string dari field released pada movies initial mejadi format datetime

>`num_mflix_comments` isinya adalah string kosong karena pada movies ada fieldnya sedangkan pada movies initial tidak ada fieldnya.

Semua diberi **conditions**(\$cond) karena jika suatu keys ada values nya maka keys tersebut akan ditampilkan, namun jika keys tersebut tidak memiliki values, maka keys tersebut tidak ditampilkan sehingga perlu digunakan **\$nonExistinField**

Pada code ini terdapat fungsi **match**() yang berguna untuk memfilter dokumen dengan hanya menghasilkan dokumen yang values imdbnya ada pada imdb_list

# Clean Movies Tunggal

In [20]:
col_tunggal = db_mov_init.clean_movies_tunggal

print('Total Documents on Clean Movies Tunggal:', col_tunggal.count())

Total Documents on Clean Movies Tunggal: 23539


***
Definisi variabel **col_tunggal** yang merupakan collection clean_movies_tunggal yang ada pada database db_mov_init. Collection tersebut merupakan hasil akhir pada case ini. Case ini telah sama keys dan valuenya dengan collection movies.
<br>
<br>Total Documents on Clean Movies Tunggal: **23539**
***

In [64]:
### Code to delete collection from database if the cluster storage full

#db_mov_init.clean_movies_tunggal.drop()

In [62]:
pprint.pprint(list(col_tunggal.find({'title': 'The House of the Devil'})))

[{'_id': ObjectId('5db008ba07645f129de5c57d'),
  'awards': {'nominations': '5',
             'text': '3 wins & 5 nominations.',
             'wins': '3'},
  'casts': ['Jocelin Donahue', 'Tom Noonan', 'Mary Woronov', 'Greta Gerwig'],
  'countries': ['USA'],
  'directors': ['Ti West'],
  'fullplot': 'College student Samantha Hughes takes a strange babysitting job '
              'that coincides with a full lunar eclipse. She slowly realizes '
              'her clients harbor a terrifying secret',
  'genres': ['Horror'],
  'imdb': {'id': 1172994, 'rating': 6.4, 'votes': 26606},
  'language': ['English'],
  'lastupdated': '2015-09-06 00:44:02.603000000',
  'metacritic': 73,
  'num_mflix_comments': '',
  'plot': 'College student Samantha Hughes takes a strange babysitting job '
          'that coincides with a full lunar eclipse. She slowly realizes her '
          'clients harbor a terrifying secret',
  'poster': 'http://ia.media-imdb.com/images/M/MV5BMTAxMDAxODg5ODReQTJeQWpwZ15BbWU3MDI5O

# 2. Validasi

## 2.1 Semua document pada clean_movies dan movie sama

In [63]:
#sorting movies collection and clean_movies_tunggal
sort_title_movies=col_mov.find({}).sort('title')
sort_title_tunggal=col_tunggal.find({}).sort('title')

In [83]:
if sort_title_movies==sort_title_tunggal:
    print('document in clean_movies_tunggal is THE SAME as documents in movies')
else:
    print('document in clean_movies_tunggal is NOT THE SAME as documents in movies')

document in clean_movies_tunggal is NOT THE SAME as documents in movies


***
Ternyata **dokumen-dokumen** yang ada pada **clean_movies_tunggal** dan **movies** **tidak sama** isinya. Hal ini dikarenakan pada movies initial pun document dengan movie yang sama dengan movies collection, isi documennya berbeda terutama pada valuesnya.
***

## 2.2 Banyak document pada clean_movies dan movie sama

In [97]:
print('Total Documents on Clean Movies Tunggal:', col_tunggal.count())
print('Total Documents on Movies              :', col_mov.count())

Total Documents on Clean Movies Tunggal: 23539
Total Documents on Movies              : 23541


***
**Banyak document** pada clean_movies dan movie **berbeda**. Selanjutnya akan dilihat apa yang menyebabkan berbeda
***

In [78]:
clean_movies = list(col_tunggal.aggregate([{"$sortByCount":"$imdb"}]))
movies = list(col_mov.aggregate([{"$sortByCount":"$imdb"}]))

for i in movies:
  if i not in clean_movies:
    print(i)

{'_id': {'id': 5, 'rating': 6.2, 'votes': 1189}, 'count': 2}


Ternyata ada 2 data pada movies dengan imdb **id : 5**, imdb **rating : 6.2** dan imdb **votes : 1189** yang tidak ada pada movies initial. Sehingga data tersebut juga tidak ada pada clean_movies_tunggal

## 2.3 Semua fields pada clean_movies ada pada movie

In [27]:
key_mov = [list(doc.keys()) for doc in col_mov.find({})]
key_mov = set([ j for i in key_mov for j in i])
key_mov

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

In [35]:
key_tunggal = [list(doc.keys()) for doc in col_tunggal.find({})]
key_tunggal = set([ j for i in key_tunggal for j in i])
key_tunggal

{'_id',
 'awards',
 'casts',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'language',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

In [41]:
print('TOTAL FIELDS MOVIES               :',len(key_mov),'\n---------------------------------------')
print('TOTAL FIELDS CLEAN MOVIES TUNGGAL :',len(key_tunggal))

TOTAL FIELDS MOVIES               : 22 
---------------------------------------
TOTAL FIELDS CLEAN MOVIES TUNGGAL : 22


***
Dapat dilihat bahwa **Semua fields pada clean_movies_tunggal ada pada movies collection**. 
<br>**Total fields** pada kedua collection tersebut adalah **22**
<br>**Jenis fields** nya **sama**
***

## 2.4 Semua value pada clean_movies sama dengan semua value pada movies dengan urutan yang sama

In [84]:
pprint.pprint(sort_title_movies[0])

{'_id': ObjectId('573a13dcf29313caabdb2734'),
 'awards': {'nominations': 1, 'text': '1 nomination.', 'wins': 0},
 'cast': ['Rhys Wakefield', 'Logan Miller', 'Ashley Hinshaw', 'Natalie Hall'],
 'countries': ['USA'],
 'directors': ['Dennis Iliadis'],
 'fullplot': 'Three college friends hit the biggest party of the year, where a '
             'mysterious phenomenon disrupts the night, quickly descending '
             'into a chaos that challenges their friendships -- and whether '
             'they can stay alive.',
 'genres': ['Sci-Fi', 'Thriller'],
 'imdb': {'id': 2395385, 'rating': 5.5, 'votes': 5626},
 'languages': ['English'],
 'lastupdated': '2015-09-11 00:09:26.053000000',
 'metacritic': 60,
 'plot': 'Three college friends hit the biggest party of the year, where a '
         'mysterious phenomenon disrupts the night, quickly descending into a '
         'chaos that challenges their friendships -- and whether they can stay '
         'alive.',
 'poster': 'https://m.media-amazon.

In [66]:
pprint.pprint(sort_title_tunggal[0])

{'_id': ObjectId('5db008cc07645f129de5e55b'),
 'awards': {'nominations': '1', 'text': '1 nomination.', 'wins': None},
 'casts': ['Rhys Wakefield', 'Logan Miller', 'Ashley Hinshaw', 'Natalie Hall'],
 'countries': ['USA'],
 'directors': ['Dennis Iliadis'],
 'fullplot': 'Three college friends hit the biggest party of the year, where a '
             'mysterious phenomenon disrupts the night, quickly descending '
             'into a chaos that challenges their friendships -- and whether '
             'they can stay alive.',
 'genres': ['Sci-Fi', 'Thriller'],
 'imdb': {'id': 2395385, 'rating': 5.5, 'votes': 5626},
 'language': ['English'],
 'lastupdated': '2015-09-11 00:09:26.053000000',
 'metacritic': 60,
 'num_mflix_comments': '',
 'plot': 'Three college friends hit the biggest party of the year, where a '
         'mysterious phenomenon disrupts the night, quickly descending into a '
         'chaos that challenges their friendships -- and whether they can stay '
         'alive.',
 'p

***
Berdasarkan Hasil pengamatan dari kedua data movies tersebut, Ada beberapa keys yang tidak sama antara clean_movies_tunggal dengan movies. untuk lebih jelasnya dapat dilihat code selanjutnya
***

### Testing Kesamaan Document Pertama

In [98]:
x = sort_title_tunggal[0]
y = sort_title_movies[0]

shared_items = {k: x[k] for k in x if k in y and x[k] == y[k]}
not_shared_items = {k: x[k] for k in x if k in y and x[k] != y[k]}

print('SAME VALUES (Total',(len(shared_items)),') :\n')
pprint.pprint(shared_items)

print('\n---------------------------------------------------------------------------------------------------------------\
      \n\nNOT SAME VALUES (Total',(len(not_shared_items)),') :\n')
pprint.pprint(not_shared_items)

SAME VALUES (Total 14 ) :

{'countries': ['USA'],
 'directors': ['Dennis Iliadis'],
 'fullplot': 'Three college friends hit the biggest party of the year, where a '
             'mysterious phenomenon disrupts the night, quickly descending '
             'into a chaos that challenges their friendships -- and whether '
             'they can stay alive.',
 'genres': ['Sci-Fi', 'Thriller'],
 'imdb': {'id': 2395385, 'rating': 5.5, 'votes': 5626},
 'lastupdated': '2015-09-11 00:09:26.053000000',
 'metacritic': 60,
 'plot': 'Three college friends hit the biggest party of the year, where a '
         'mysterious phenomenon disrupts the night, quickly descending into a '
         'chaos that challenges their friendships -- and whether they can stay '
         'alive.',
 'rated': 'NOT RATED',
 'released': datetime.datetime(2013, 10, 24, 0, 0),
 'title': 1,
 'type': 'movie',
 'writers': ['Dennis Iliadis (story)', 'Bill Gullo (screenplay)'],
 'year': 2013}

--------------------------------------

***
Berdasarkan hasil code diatas, dapat dilihat dengan lebih jelas bahwa **Tidak semua values** pada **clean_movies_tunggal** dan **movies** adalah sama baik urutan maupun valuesnya. Terdapat perbedaan values dan urutan values pada kedua collection tersebut. Lagi-Lagi Hal ini disebabkan karena pada movies initial pun document dengan movie yang sama dengan movies collection, isi documennya berbeda terutama pada valuesnya.
<br>
<br>Pada code diatas, diambil **sampel document** clean_movies_tunggal dan movies dengan urutan **pertama** dengan judul movies 1. Dapat dilihat bahwa pada document pertama tesebut terdapat **5 keys dengan values yang berbeda dan urutan yang berbeda**.
***

### Testing Kesamaan 100 Document Pertama

In [109]:
x2 = sort_title_tunggal
y2 = sort_title_movies

total_shared_items = 0
total_not_shared_items = 0

for i in range(100) :
    #same values
    shared_items2 = {k: x2[i][k] for k in x2[i] if k in y2[i] and x2[i][k] == y2[i][k]}
    total_shared_items += len(shared_items2)
    #not same values
    not_shared_items2 = {k: x2[i][k] for k in x2[i] if k in y2[i] and x2[i][k] != y2[i][k]}
    total_not_shared_items += len(not_shared_items2)

In [113]:
print('SAME VALUES Total     :',(total_shared_items))

print('----------------------------\
      \nNOT SAME VALUES Total :',(total_not_shared_items))

SAME VALUES Total     : 1262
----------------------------      
NOT SAME VALUES Total : 558


***
Untuk Mengecek kesamaan semua values pada kedua collection, dibutuhkan waktu yang lama karena documentnya yang banyak yakni kurang lebih 23 ribuan data. Sehingga untuk mengecek kesamaan semuanya, dibutuhkan waktu yang lama. Oleh karena itu kali ini hanya menggunakan 100 document sebagai sampel.
<br>Dari hasil diatas dapat dilihat bahwa pada **100 document pertama** pada **clean_movies_tunggal**, values yang berbeda dengan movies adalah **558**, sedangkan values yang sama dengan movies adalah **1262**
***

# KESIMPULAN :
1. Total data Movies adalah **23541** dan total data Movies Initial adalah **46014**, sehingga untuk membuat Clean Movies, diperlukan proses agregasi untuk menghasilkan data yang sama dengan data movies menggunakan **pipeline**. 
2. untuk melakukan menentukan keys dan mengisi values yang akan ada pada collection baru pada mongodb, digunakan pipeline stage agregasi dengan **\$project** 
3. untuk memfilter dokumen-dokumen yang akan ada pada collection baru pada mongodb, digunakan pipeline stage agregasi dengan **\$match**
4. untuk menamai collection pada collection baru yang akan dibuat, digunakan fungsi pipeline stage agregasi **\$out**
5. Hasilnya (Clean Movies) **tidak 100\% sama** dengan Movies.Jumlah document antar kedua collection tersebut juga berselisih 2. Total document pada Clean Movies adalah **23539**. Kemudian untuk jumlah field dan jenis fieldnya sudah sama. Namun masih banyak isi(values) yang berbeda pada setiap documentnya. Hal ini dikarenakan pada movies initial pun dokumen dengan movie yang sama dengan movies collection, isi dokumennya berbeda terutama pada valuesnya. Namun sebagian besar document antar kedua fields tesebut sudah sama.